### PR TITLE
Support high-resolution sampling on ICM4xxxx IMUs

### PR DESCRIPTION
**Pull request by andyp1per**
ICM42688, IIM42652, ICM42670, ICM45686 support higher resolution gyro and accel data in the FIFO. This higher resolution is particularly useful on large range devices (ICM45686) and prevents the quantization that would otherwise occur with these devices. 

High-resolution support is enabled via ```INS_HIRES_SAMPLE``` which is a mask that can be used in a similar way to ```INS_FAST_SAMPLE```

This also fixes a bug with filter setup where we were relying on device ids that had not yet been set.

This work was sponsored by CubePIlot

### CODE REVIEW

**Comment by andyp1per** [comment]
Flown on my 5" quad no issues

**Comment by tridge** [comment]
@andyp1per I'd like a discuss a bit, particularly regarding testing on all the v3 variants, and if we need the parameter or should just always enable

**Review by tridge — CHANGES_REQUESTED** [review]
a few changes:
 - needs to be under an #if, enabled in hwdef.dat
 - needs careful testing on all supported IMUs
 - need to test with raw 32kHz fifo logging to ensure this is really a benefit, with logging of the raw 32bit signed number
 - we need to check if enabling high res changes gyro or accel offsets

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp [hunk 1]
```diff
@@ -247,15 +267,53 @@ void AP_InertialSensor_Invensensev3::start()
     case Invensensev3_Type::ICM45686:
         devtype = DEVTYPE_INS_ICM45686;
         temp_sensitivity = 1.0 / 2.0;
+        gyro_scale = GYRO_SCALE_4000DPS;
+        accel_scale = ACCEL_SCALE_32G;
         break;
     case Invensensev3_Type::ICM40609:
     default:
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp:279]: better without the default

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp [hunk 2]
```diff
@@ -747,8 +896,6 @@ bool AP_InertialSensor_Invensensev3::check_whoami(void)
     switch (whoami) {
     case INV3_ID_ICM40609:
         inv3_type = Invensensev3_Type::ICM40609;
-        // Accel scale 32g (1024 LSB/g)
-        accel_scale = (GRAVITY_MSS / 1024);
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp:751]: this looks like it changes accel_scale for 40609

**Review by tridge — CHANGES_REQUESTED** [review]
moving to an enum class will help

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.h [hunk 3]
```diff
@@ -86,20 +87,44 @@ class AP_InertialSensor_Invensensev3 : public AP_InertialSensor_Backend
...
+    static constexpr float ACCEL_SCALE_HIGHRES_16G = (GRAVITY_MSS / (SCALE_RANGE_20BIT / 16));
+    static constexpr float ACCEL_SCALE_HIGHRES_32G = (GRAVITY_MSS / (SCALE_RANGE_20BIT / 32));
 
-    float accel_scale = (GRAVITY_MSS / 2048);
+    float accel_scale = ACCEL_SCALE_16G;
     float gyro_scale = GYRO_SCALE_2000DPS;
 
     // are we doing more than 1kHz sampling?
     bool fast_sampling;
+    bool highres_sampling;
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.h:128]: Fixed

**Review by tridge — APPROVED** [review]

**Pull request merged**